gdal/frmts/netcdf/netcdflayer.cpp

```diff
@@ -2508,51 +2813,87 @@ OGRErr netCDFLayer::CreateField(OGRFieldDefn *poFieldDefn, int /* bApproxOK */)
     m_aoFieldDesc.push_back(fieldDesc);
 
     const char *pszLongName = CPLSPrintf("Field %s", poFieldDefn->GetNameRef());
-    status = nc_put_att_text(m_nLayerCDFId, nVarID, CF_LNG_NAME,
-                             strlen(pszLongName), pszLongName);
-    NCDF_ERR(status);
 
     if ( !m_bLegacyCreateMode )
     {
-        char ct_name_carr[NC_MAX_NAME + 1] = {0};
-        status = nc_inq_varname(m_nLayerCDFId, m_writableSGContVarID, ct_name_carr);
-        NCDF_ERR(status);
+        m_poDS->vcdf.nc_put_vatt_text(nVarID, CF_LNG_NAME, pszLongName);
+
+        std::string ct_name(m_layerSGDefn.get_containerName());
+        m_poDS->vcdf.nc_put_vatt_text(nVarID, CF_SG_GEOMETRY, ct_name.c_str());
+    }
 
-        status = nc_put_att_text(m_nLayerCDFId, nVarID, CF_SG_GEOMETRY, strlen(ct_name_carr), ct_name_carr);
+    else
+    {
+        status = nc_put_att_text(m_nLayerCDFId, nVarID, CF_LNG_NAME,
+                                 strlen(pszLongName), pszLongName);
         NCDF_ERR(status);
     }
 
     if( m_bWriteGDALTags )
     {
-        status = nc_put_att_text(m_nLayerCDFId, nVarID, "ogr_field_name",
-                                 strlen(poFieldDefn->GetNameRef()),
-                                 poFieldDefn->GetNameRef());
-        NCDF_ERR(status);
+        if ( !m_bLegacyCreateMode )
+        {
+            m_poDS->vcdf.nc_put_vatt_text(nVarID, "ogr_field_name", poFieldDefn->GetNameRef());
+        }
+
+        else
+        {
+            status = nc_put_att_text(m_nLayerCDFId, nVarID, "ogr_field_name",
+                                     strlen(poFieldDefn->GetNameRef()),
+                                     poFieldDefn->GetNameRef());
+            NCDF_ERR(status);
+        }
 
         const char *pszType = OGRFieldDefn::GetFieldTypeName(eType);
         if( eSubType != OFSTNone )
         {
             pszType = CPLSPrintf("%s(%s)", pszType,
                                  OGRFieldDefn::GetFieldSubTypeName(eSubType));
         }
-        status = nc_put_att_text(m_nLayerCDFId, nVarID, "ogr_field_type",
-                                 strlen(pszType), pszType);
-        NCDF_ERR(status);
+
+        if ( !m_bLegacyCreateMode )
+        {
+            m_poDS->vcdf.nc_put_vatt_text(nVarID, "ogr_field_type", pszType);
+        }
+
+        else
+        {
+            status = nc_put_att_text(m_nLayerCDFId, nVarID, "ogr_field_type",
+                                     strlen(pszType), pszType);
+            NCDF_ERR(status);
+        }
 
         const int nWidth = poFieldDefn->GetWidth();
         if(nWidth || nType == NC_CHAR)
         {
-            status = nc_put_att_int(m_nLayerCDFId, nVarID, "ogr_field_width",
-                                    NC_INT, 1, &nWidth);
-            NCDF_ERR(status);
+            if ( !m_bLegacyCreateMode )
```

Comment:
see above remarks. could be merged in a single code path with a passthrough vcdf implementation